src/lib/agent.c

```diff
@@ -36,58 +37,59 @@
 
 struct sa *sa;
 
-/*
- * Notify the SA that we are about to enter idle state (WFI).
- */
-void sa_enter_idle(struct sof *sof)
-{
-	struct sa *sa = sof->sa;
-
-	sa->last_idle = platform_timer_get(platform_timer);
-}
-
 static enum task_state validate(void *data)
 {
 	struct sa *sa = data;
 	uint64_t current;
 	uint64_t delta;
 
-	current = platform_timer_get(platform_timer);
-	delta = current - sa->last_idle;
+	if (sa->is_active) {
+		current = platform_timer_get(platform_timer);
+		delta = current - sa->last_check;
 
-	/* were we last idle longer than timeout */
+		/* panic timeout */
+		if (delta > sa->panic_timeout)
+			panic(SOF_IPC_PANIC_IDLE);
 
-	if (delta > sa->ticks && sa->is_active) {
-		trace_sa("validate(), idle longer than timeout, delta = %u",
-			delta);
-		panic(SOF_IPC_PANIC_IDLE);
+		/* warning timeout */
+		if (delta > sa->warn_timeout)
+			trace_sa_error("validate(), ll drift detected, delta = "
+				       "%u", delta);
```

Comment:
What about logging a time of the delay in microseconds instead of bare delta? I am afraid during debugging delta won't tell us much and we will end up calculating it manually anyway.